File: THOUGHT/LAB/CAT_CAS/7_decoder/50_phase_bm_cpu/50_6_fixed_point_substrate/14_noncollapse_frontier/phase6b6/live_small_wall/audio_frequency_wave_substrate/cat_cas_phase_frontier/fixed_resident_torus.c
/*
 * Mutable CAT_CAS frontier: direct C implementation of the fixed-resident
 * reversible torus. This is ordinary CPU execution used to remove Python
 * interpreter overhead and measure the mechanism honestly on the lab target.
 *
 * Native evolution carries only complex relations. Discrete root decoding
 * occurs after the result latch survives the inverse traversal.
 */

#define _GNU_SOURCE
#include <complex.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define M 31
#define K 5
#define CHANNELS 2
#define CELLS (K * CHANNELS * M)
#define CRT_MODULUS 15015
#define RESTORATION_MAX 2.0e-11
#define PI 3.141592653589793238462643383279502884

static const int phase_moduli[K] = {3, 5, 7, 11, 13};

static inline size_t cell_index(int phase, int channel, int residue) {
    return (size_t)((phase * CHANNELS + channel) * M + residue);
}

static inline double complex unit(double complex value) {
    const double magnitude = cabs(value);
    if (!(magnitude > 0.0) || !isfinite(magnitude)) {
        fprintf(stderr, "collapsed or nonfinite phase relation\n");
        exit(2);
    }
    return value / magnitude;
}

static inline double complex relation_at(
    const double complex *carrier, int phase, int residue
) {
    const double complex reference =
        carrier[cell_index(phase, 0, residue)];
    const double complex signal =
        carrier[cell_index(phase, 1, residue)];
    return unit(signal * conj(reference));
}

static inline double complex complex_power_int(
    double complex base, int exponent
) {
    double complex result = 1.0 + 0.0 * I;
    double complex factor = base;
    int remaining = exponent;
    while (remaining > 0) {
        if (remaining & 1) {
            result *= factor;
        }
        factor *= factor;
        remaining >>= 1;
    }
    return result;
}

static inline double complex phase_lock(
    double complex value, int modulus
) {
    double complex locked = unit(value);
    const double error_force =
        cimag(complex_power_int(locked, modulus));
    const double correction = error_force / (double)modulus;
    locked *= 1.0 - I * correction;
    return unit(locked);
}

static int mod_inverse(int value, int modulus) {
    for (int candidate = 1; candidate < modulus; ++candidate) {
        if ((value * candidate) % modulus == 1) {
            return candidate;
        }
    }
    fprintf(stderr, "missing modular inverse\n");
    exit(2);
}

static int public_weight(uint64_t index, int family) {
    const uint64_t linear = (uint64_t)(17 + 2 * family);
    const uint64_t quadratic = (uint64_t)(5 + family);
    return 1 + (int)(
        (index * linear + index * index * quadratic + 7U + 11U * family)
        % (M - 1)
    );
}

static void make_borrowed(
    double complex *carrier, int identity
) {
    for (int phase = 0; phase < K; ++phase) {
        for (int residue = 0; residue < M; ++residue) {
            const double angle =
                0.173
                + 0.071 * phase
                + 0.037 * residue
                + 0.023 * sin(0.17 * residue + identity * 0.07);
            const double amplitude =
                0.71
                + 0.08 * cos(0.13 * residue + 0.05 * phase);
            const double complex rail = amplitude * cexp(I * angle);
            for (int channel = 0; channel < CHANNELS; ++channel) {
                carrier[cell_index(phase, channel, residue)] =
                    rail;
            }
        }
    }
}

static void seed_carrier(double complex *carrier) {
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        for (int residue = 0; residue < M; ++residue) {
            const double complex initial =
                residue == 0
                ? cexp(2.0 * PI * I / (double)modulus)
                : 1.0 + 0.0 * I;
            const double complex before =
                relation_at(carrier, phase, residue);
            if (cabs(before - 1.0) > 1.0e-12) {
                fprintf(stderr, "borrowed twin rails lost common mode\n");
                exit(2);
            }
            carrier[cell_index(phase, 1, residue)] *= initial;
        }
    }
}

static void forward_shift(double complex *carrier, int shift) {
    double complex before[K][M];
    double complex after[K][M];
    for (int phase = 0; phase < K; ++phase) {
        for (int residue = 0; residue < M; ++residue) {
            before[phase][residue] =
                relation_at(carrier, phase, residue);
        }
    }
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        for (int residue = 0; residue < M; ++residue) {
            const int shifted = (residue - shift + M) % M;
            after[phase][residue] = phase_lock(
                before[phase][residue] * before[phase][shifted],
                modulus
            );
        }
    }
    for (int phase = 0; phase < K; ++phase) {
        for (int residue = 0; residue < M; ++residue) {
            carrier[cell_index(phase, 1, residue)] =
                after[phase][residue]
                * carrier[cell_index(phase, 0, residue)];
        }
    }
}

static void inverse_shift(double complex *carrier, int shift) {
    double complex before[K][M];
    double complex after[K][M];
    for (int phase = 0; phase < K; ++phase) {
        for (int residue = 0; residue < M; ++residue) {
            before[phase][residue] =
                relation_at(carrier, phase, residue);
        }
    }
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        int positions[M];
        for (int index = 0; index < M; ++index) {
            positions[index] = (index * shift) % M;
        }
        double complex alternating = 1.0 + 0.0 * I;
        for (int index = 0; index < M; ++index) {
            const double complex value = before[phase][positions[index]];
            alternating *= (index == 0 || (index & 1))
                ? value
                : conj(value);
        }
        after[phase][positions[0]] = phase_lock(
            complex_power_int(
                alternating,
                mod_inverse(2, modulus)
            ),
            modulus
        );
        for (int index = 1; index < M; ++index) {
            const int current = positions[index];
            const int previous = positions[index - 1];
            after[phase][current] = phase_lock(
                before[phase][current] * conj(after[phase][previous]),
                modulus
            );
        }
    }
    for (int phase = 0; phase < K; ++phase) {
        for (int residue = 0; residue < M; ++residue) {
            carrier[cell_index(phase, 1, residue)] =
                after[phase][residue]
                * carrier[cell_index(phase, 0, residue)];
        }
    }
}

static int decode_latch(
    const double complex latch[K],
    double *maximum_root_distance
) {
    int residues[K];
    *maximum_root_distance = 0.0;
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        const double complex value = unit(latch[phase]);
        double best = INFINITY;
        int best_index = -1;
        for (int candidate = 0; candidate < modulus; ++candidate) {
            const double complex root =
                cexp(2.0 * PI * I * candidate / (double)modulus);
            const double distance = cabs(root - value);
            if (distance < best) {
                best = distance;
                best_index = candidate;
            }
        }
        residues[phase] = best_index;
        if (best > *maximum_root_distance) {
            *maximum_root_distance = best;
        }
    }
    int64_t total = 0;
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        const int partial = CRT_MODULUS / modulus;
        total += (int64_t)residues[phase]
            * partial
            * mod_inverse(partial % modulus, modulus);
    }
    return (int)(total % CRT_MODULUS);
}

static uint64_t monotonic_ns(void) {
    struct timespec value;
    if (clock_gettime(CLOCK_MONOTONIC_RAW, &value) != 0) {
        perror("clock_gettime");
        exit(2);
    }
    return (uint64_t)value.tv_sec * UINT64_C(1000000000)
        + (uint64_t)value.tv_nsec;
}

static double maximum_error(
    const double complex *left, const double complex *right
) {
    double maximum = 0.0;
    for (size_t index = 0; index < CELLS; ++index) {
        const double error = cabs(left[index] - right[index]);
        if (error > maximum) {
            maximum = error;
        }
    }
    return maximum;
}

static double displacement_l2(
    const double complex *left, const double complex *right
) {
    double sum = 0.0;
    for (size_t index = 0; index < CELLS; ++index) {
        const double error = cabs(left[index] - right[index]);
        sum += error * error;
    }
    return sqrt(sum);
}

static void run_case(
    double complex *borrowed,
    uint64_t steps,
    int family,
    int target
) {
    double complex working[CELLS];
    double complex latch[K];
    memcpy(working, borrowed, sizeof(working));

    const uint64_t phase_start = monotonic_ns();
    seed_carrier(working);
    for (uint64_t index = 0; index < steps; ++index) {
        forward_shift(working, public_weight(index, family));
    }
    const double displacement = displacement_l2(working, borrowed);
    for (int phase = 0; phase < K; ++phase) {
        latch[phase] = relation_at(working, phase, target);
    }
    for (uint64_t index = steps; index-- > 0;) {
        inverse_shift(working, public_weight(index, family));
    }
    for (int phase = 0; phase < K; ++phase) {
        const int modulus = phase_moduli[phase];
        for (int residue = 0; residue < M; ++residue) {
            const double complex initial =
                residue == 0
                ? cexp(2.0 * PI * I / (double)modulus)
                : 1.0 + 0.0 * I;
            working[cell_index(phase, 1, residue)] *=
                conj(initial);
        }
    }
    const uint64_t phase_ns = monotonic_ns() - phase_start;
    const double restoration = maximum_error(working, borrowed);

    double root_distance = 0.0;
    const int phase_count = decode_latch(latch, &root_distance);

    printf(
        "{\"steps\":%llu,\"family\":%d,\"phase_count\":%d,"
        "\"phase_ns\":%llu,"
        "\"resident_complex_cells\":%d,\"displacement_l2\":%.12g,"
        "\"restoration_max_abs\":%.12g,\"root_distance_max\":%.12g,"
        "\"restoration_pass\":%s}\n",
        (unsigned long long)steps,
        family,
        phase_count,
        (unsigned long long)phase_ns,
        CELLS,
        displacement,
        restoration,
        root_distance,
        restoration <= RESTORATION_MAX ? "true" : "false"
    );
    memcpy(borrowed, working, sizeof(working));
}

int main(int argc, char **argv) {
    const uint64_t sizes[] = {
        16, 64, 256, 1024, 4096, 16384, 65536
    };
    double complex carrier[CELLS];
    make_borrowed(carrier, 77);
    if (argc > 1) {
        char *end = NULL;
        const unsigned long long requested = strtoull(argv[1], &end, 10);
        if (
            end == argv[1]
            || *end != '\0'
            || requested == 0
        ) {
            fprintf(stderr, "step count must be a positive integer\n");
            return 2;
        }
        const int family = argc > 2 ? atoi(argv[2]) : 0;
        if (family < 0 || family > 1) {
            fprintf(stderr, "family must be 0 or 1\n");
            return 2;
        }
        run_case(carrier, (uint64_t)requested, family, 17);
        return 0;
    }
    for (int family = 0; family < 2; ++family) {
        for (size_t index = 0; index < sizeof(sizes) / sizeof(sizes[0]); ++index) {
            run_case(carrier, sizes[index], family, 17);
        }
    }
    return 0;
}

File: THOUGHT/LAB/CAT_CAS/7_decoder/50_phase_bm_cpu/50_6_fixed_point_substrate/14_noncollapse_frontier/phase6b6/live_small_wall/audio_frequency_wave_substrate/cat_cas_phase_frontier/fixed_resident_torus.py
"""Mutable CAT_CAS frontier: fixed-resident reversible phase path process.

This is development code, not a frozen evidence package.

For an odd prime cyclic width M and nonzero shift w, the simultaneous phase
map

    y[s] = x[s] * x[s-w]

is a bijection on p-th-root relations for every odd phase modulus p.  The map
therefore needs one resident torus rather than a fresh O(M) layer per program
step.  Its inverse is derived from the same public shift and performs only
complex phase products, conjugations, and fixed integer phase powers.

The sole discrete conversion remains ``collapse_boundary``.
"""

from __future__ import annotations

import ast
import hashlib
import json
import math
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
from typing import Any, Iterable

import numpy as np


ENGINE_VERSION = "0.1.0-development"
DEFAULT_PHASE_MODULI = (3, 5, 7, 11, 13)
ROOT_DISTANCE_MAX = 2.0e-9
ROOT_MARGIN_MIN = 0.15
RESTORATION_MAX = 2.0e-11
PHASE_LOCK_ITERATIONS = 1


def canonical_bytes(document: Any) -> bytes:
    return (
        json.dumps(
            document,
            allow_nan=False,
            ensure_ascii=True,
            separators=(",", ":"),
            sort_keys=True,
        )
        + "\n"
    ).encode("utf-8")


def sha256_bytes(payload: bytes) -> str:
    return hashlib.sha256(payload).hexdigest()


def _is_prime(value: int) -> bool:
    if value < 2:
        return False
    if value % 2 == 0:
        return value == 2
    return all(value % divisor for divisor in range(3, math.isqrt(value) + 1, 2))


def _pairwise_coprime(values: Iterable[int]) -> bool:
    items = tuple(values)
    return all(
        math.gcd(items[left], items[right]) == 1
        for left in range(len(items))
        for right in range(left + 1, len(items))
    )


@dataclass(frozen=True)
class PhaseProgram:
    name: str
    residue_modulus: int
    phase_moduli: tuple[int, ...]
    weights: tuple[int, ...]
    target_residue: int

    def validate(self) -> None:
        if not self.name or len(self.name) > 128:
            raise ValueError("program name is missing or too long")
        if not _is_prime(self.residue_modulus) or self.residue_modulus == 2:
            raise ValueError("residue modulus must be an odd prime")
        if not self.phase_moduli:
            raise ValueError("phase moduli are required")
        if (
            any(not _is_prime(value) or value == 2 for value in self.phase_moduli)
            or not _pairwise_coprime(self.phase_moduli)
        ):
            raise ValueError("phase moduli must be pairwise-coprime odd primes")
        if not self.weights:
            raise ValueError("at least one public shift is required")
        if any(
            not isinstance(weight, int)
            or not 0 < weight < self.residue_modulus
            for weight in self.weights
        ):
            raise ValueError("every public shift must be nonzero and in range")
        if not 0 <= self.target_residue < self.residue_modulus:
            raise ValueError("target residue is outside the torus")

    def document(self) -> dict[str, Any]:
        return {
            "geometry": {
                "phase_moduli": list(self.phase_moduli),
                "residue_modulus": self.residue_modulus,
            },
            "name": self.name,
            "public_shifts": list(self.weights),
            "target_residue": self.target_residue,
        }

    @property
    def fingerprint(self) -> str:
        self.validate()
        return sha256_bytes(canonical_bytes(self.document()))


def carrier_shape(program: PhaseProgram) -> tuple[int, int, int]:
    program.validate()
    return len(program.phase_moduli), 2, program.residue_modulus


def borrowed_carrier(program: PhaseProgram, identity: int = 0) -> np.ndarray:
    """Construct a deterministic noncanonical carrier for development probes."""

    phase_count, channel_count, residue_count = carrier_shape(program)
    p = np.arange(phase_count, dtype=np.float64)[:, None, None]
    residue = np.arange(residue_count, dtype=np.float64)[None, None, :]
    phase = (
        0.173
        + 0.071 * p
        + 0.037 * residue
        + 0.023 * np.sin(0.17 * residue + identity * 0.07)
    )
    amplitude = (
        0.71
        + 0.08 * np.cos(0.13 * residue + 0.05 * p)
    )
    rail = amplitude * np.exp(1j * phase)
    result = np.repeat(rail, channel_count, axis=1)
    return np.asarray(result, dtype=np.complex128)


def _as_carrier(program: PhaseProgram, carrier: np.ndarray) -> np.ndarray:
    candidate = np.asarray(carrier, dtype=np.complex128)
    if candidate.shape != carrier_shape(program):
        raise ValueError("borrowed carrier has the wrong fixed geometry")
    if (
        not np.all(np.isfinite(candidate))
        or float(np.min(np.abs(candidate))) <= 1.0e-14
    ):
        raise ValueError("borrowed carrier must be finite and nonzero")
    return np.array(candidate, copy=True)


def _relations(carrier: np.ndarray) -> np.ndarray:
    relation = carrier[:, 1, :] * np.conjugate(carrier[:, 0, :])
    magnitude = np.abs(relation)
    if float(np.min(magnitude)) <= 1.0e-15:
        raise ValueError("carrier relation collapsed")
    return relation / magnitude


def _initial_relations(program: PhaseProgram) -> np.ndarray:
    result = np.ones(
        (len(program.phase_moduli), program.residue_modulus),
        dtype=np.complex128,
    )
    for phase_index, modulus in enumerate(program.phase_moduli):
        result[phase_index, 0] = np.exp(2j * math.pi / modulus)
    return result


def _seed(program: PhaseProgram, carrier: np.ndarray) -> np.ndarray:
    working = _as_carrier(program, carrier)
    if float(np.max(np.abs(_relations(working) - 1.0))) > 1.0e-12:
        raise ValueError("borrowed twin rails lost their common-mode relation")
    working[:, 1, :] *= _initial_relations(program)
    return working


def _lock(row: np.ndarray, modulus: int) -> np.ndarray:
    """Fixed p-fold phase dynamic; it does not label or select a root."""

    locked = row / np.abs(row)
    for _ in range(PHASE_LOCK_ITERATIONS):
        locked *= np.exp(-1j * np.imag(locked**modulus) / float(modulus))
        locked /= np.abs(locked)
    return locked


def _lock_rows(
    rows: np.ndarray, moduli: tuple[int, ...]
) -> np.ndarray:
    locked = rows / np.abs(rows)
    powers = np.asarray(moduli, dtype=np.int64)[:, None]
    divisors = np.asarray(moduli, dtype=np.float64)[:, None]
    for _ in range(PHASE_LOCK_ITERATIONS):
        locked *= np.exp(-1j * np.imag(locked**powers) / divisors)
        locked /= np.abs(locked)
    return locked


def _forward_relation(
    relation: np.ndarray, shift: int, modulus: int
) -> np.ndarray:
    return _lock(relation * np.roll(relation, shift), modulus)


@lru_cache(maxsize=None)
def _inverse_kernel(width: int, shift: int, modulus: int) -> np.ndarray:
    """Return public circulant inverse coefficients for I + P^w."""

    inverse_two = pow(2, -1, modulus)
    kernel = np.zeros((width, width), dtype=np.int64)
    for output in range(width):
        for distance in range(width):
            source = (output - distance * shift) % width
            sign = 1 if distance % 2 == 0 else -1
            kernel[output, source] = (inverse_two * sign) % modulus
    kernel.setflags(write=False)
    return kernel


@lru_cache(maxsize=None)
def _inverse_kernels(
    width: int, shift: int, moduli: tuple[int, ...]
) -> np.ndarray:
    kernels = np.stack(
        tuple(
            _inverse_kernel(width, shift, modulus)
            for modulus in moduli
        ),
        axis=0,
    )
    kernels.setflags(write=False)
    return kernels


def _inverse_relation(
    relation: np.ndarray, shift: int, modulus: int
) -> np.ndarray:
    """Invert y[s] = x[s] x[s-w] on one odd cyclic phase group.

    Along the unique cycle ``0,w,2w,...`` the inverse is

        x = (1/2) * (I - P^w + P^(2w) - ... + P^((M-1)w)) * y.

    The fixed public coefficient network is applied as complex phase powers and
    products. It never converts a relation to an exponent label.
    """

    width = relation.size
    kernel = _inverse_kernel(width, shift, modulus)
    recovered = np.prod(relation[None, :] ** kernel, axis=1)
    return _lock(recovered, modulus)


def _forward(carrier: np.ndarray, shift: int, moduli: tuple[int, ...]) -> None:
    before = _relations(carrier)
    after = _lock_rows(
        before * np.roll(before, shift, axis=1),
        moduli,
    )
    carrier[:, 1, :] = after * carrier[:, 0, :]


def _inverse(carrier: np.ndarray, shift: int, moduli: tuple[int, ...]) -> None:
    before = _relations(carrier)
    kernels = _inverse_kernels(before.shape[1], shift, moduli)
    after = _lock_rows(
        np.prod(before[:, None, :] ** kernels, axis=2),
        moduli,
    )
    carrier[:, 1, :] = after * carrier[:, 0, :]


@dataclass(frozen=True)
class PhaseExecution:
    program: PhaseProgram
    result_latch: np.ndarray
    restored_carrier: np.ndarray
    displacement_l2: float
    restoration_max_abs: float
    resident_complex_cells: int
    public_program_words: int
    history_factor_count: int
    inverse_mode: str


def execute_catalytic(
    program: PhaseProgram,
    carrier: np.ndarray,
    *,
    forward_mode: str = "correct",
    inverse_mode: str = "correct",
) -> PhaseExecution:
    """Borrow, evolve, latch, reverse, and restore one fixed torus."""

    program.validate()
    borrowed = _as_carrier(program, carrier)
    working = _seed(program, borrowed)
    if forward_mode == "correct":
        forward_shifts = program.weights
    elif forward_mode == "remove_last":
        forward_shifts = program.weights[:-1]
    elif forward_mode == "scramble_geometry":
        forward_shifts = tuple(
            (weight * weight + 1) % program.residue_modulus or 1
            for weight in program.weights
        )
    elif forward_mode == "removed":
        forward_shifts = ()
    else:
        raise ValueError("unknown forward mode")

    for shift in forward_shifts:
        _forward(working, shift, program.phase_moduli)
    displacement = float(np.linalg.norm(working - borrowed))
    latch = np.array(
        _relations(working)[:, program.target_residue], copy=True
    )

    if inverse_mode == "correct":
        inverse_shifts = tuple(reversed(forward_shifts))
    elif inverse_mode == "wrong_program":
        inverse_shifts = tuple(
            reversed(
                tuple(
                    (shift + 1) % program.residue_modulus or 1
                    for shift in forward_shifts
                )
            )
        )
    elif inverse_mode == "omitted":
        inverse_shifts = ()
    else:
        raise ValueError("unknown inverse mode")
    for shift in inverse_shifts:
        _inverse(working, shift, program.phase_moduli)
    if inverse_mode != "omitted":
        working[:, 1, :] *= np.conjugate(_initial_relations(program))

    return PhaseExecution(
        program=program,
        result_latch=latch,
        restored_carrier=working,
        displacement_l2=displacement,
        restoration_max_abs=float(np.max(np.abs(working - borrowed))),
        resident_complex_cells=int(working.size),
        public_program_words=len(program.weights),
        history_factor_count=0,
        inverse_mode=inverse_mode,
    )


@dataclass(frozen=True)
class BoundaryResult:
    valid: bool
    residues: tuple[int, ...] | None
    count_mod_crt: int | None
    root_distances: tuple[float, ...]
    root_margins: tuple[float, ...]


def _crt(residues: tuple[int, ...], moduli: tuple[int, ...]) -> int:
    product = math.prod(moduli)
    total = 0
    for residue, modulus in zip(residues, moduli):
        partial = product // modulus
        total += residue * partial * pow(partial, -1, modulus)
    return total % product


def collapse_boundary(
    program: PhaseProgram, result_latch: np.ndarray
) -> BoundaryResult:
    """The sole phase-to-discrete conversion."""

    latch = np.asarray(result_latch, dtype=np.complex128)
    if latch.shape != (len(program.phase_moduli),):
        raise ValueError("result latch has the wrong geometry")
    residues: list[int] = []
    distances: list[float] = []
    margins: list[float] = []
    for value, modulus in zip(latch, program.phase_moduli):
        phasor = value / abs(value)
        roots = np.exp(2j * math.pi * np.arange(modulus) / modulus)
        root_distances = np.abs(roots - phasor)
        order = np.argsort(root_distances, kind="stable")
        residues.append(int(order[0]))
        distances.append(float(root_distances[order[0]]))
        margins.append(
            float(root_distances[order[1]] - root_distances[order[0]])
        )
    valid = bool(
        max(distances, default=math.inf) <= ROOT_DISTANCE_MAX
        and min(margins, default=-math.inf) >= ROOT_MARGIN_MIN
    )
    decoded = tuple(residues) if valid else None
    return BoundaryResult(
        valid=valid,
        residues=decoded,
        count_mod_crt=(
            None if decoded is None else _crt(decoded, program.phase_moduli)
        ),
        root_distances=tuple(distances),
        root_margins=tuple(margins),
    )


def maximum_abs_error(left: np.ndarray, right: np.ndarray) -> float:
    return float(
        np.max(
            np.abs(
                np.asarray(left, dtype=np.complex128)
                - np.asarray(right, dtype=np.complex128)
            )
        )
    )


def source_no_smuggle() -> dict[str, Any]:
    source = Path(__file__).read_text(encoding="utf-8")
    tree = ast.parse(source)
    native_names = {
        "_forward_relation",
        "_inverse_relation",
        "_forward",
        "_inverse",
        "execute_catalytic",
    }
    forbidden = {
        "argmax",
        "expected_result",
        "oracle",
        "solve",
        "enumerate_candidates",
    }
    calls: set[str] = set()
    for node in tree.body:
        if isinstance(node, ast.FunctionDef) and node.name in native_names:
            for child in ast.walk(node):
                if isinstance(child, ast.Call):
                    if isinstance(child.func, ast.Name):
                        calls.add(child.func.id)
                    elif isinstance(child.func, ast.Attribute):
                        calls.add(child.func.attr)
    hits = sorted(calls & forbidden)
    return {
        "complete_path_modes": 0,
        "decoded_feedback": False,
        "forbidden_call_hits": hits,
        "history_factor_count": 0,
        "passed": not hits,
    }


def engine_fingerprint() -> str:
    return sha256_bytes(
        canonical_bytes(
            {
                "engine_version": ENGINE_VERSION,
                "source_sha256": sha256_bytes(Path(__file__).read_bytes()),
            }
        )
    )


if not source_no_smuggle()["passed"]:
    raise RuntimeError("fixed-resident phase engine contains a forbidden path")
